Add many forward grad

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43718 make jvp use forward mode by default
* #43717 Add gradcheck for forward grad
* **#43716 Add many forward grad**
* #43715 Basic auto formula for forward grad
* #43714 make sure SavedVariable propagate fw_grad properly
* #43713 Add forward grad mode in python and c++
* #43712 Add fw_grad field and accessors

This PR add some helper to generate forward mode formula automatically based on the forward and backward formula.
The special case that are interesting to re-use are:
- if the function is element-wise (so there is a 1-1 mapping between the elements of the input and output), then the Jacobian is diagonal. And in this case, J v and v^T J are actually the same thing. So we can just re-use the backward function as-is, just providing a different v.
- if the function is linear (not affine!) of a single input (we might be able to do more than one, but that’s tricky) then f(x) = A x for some matrix A. And the Jacobian of f is also A. So we can do J v = A v = f(v). So we can just call the original function, replacing the differentiable input with the vector v of interest.

Differential Revision: [D23454148](https://our.internmc.facebook.com/intern/diff/D23454148)

albanD committed:
Add many forward grad

[ghstack-poisoned]

albanD committed:
Update on "Add many forward grad"




[ghstack-poisoned]

zou3519 commented on tools/autograd/derivatives.yaml:
```diff
@@ -164,13 +164,16 @@
 
 - name: acos(Tensor self) -> Tensor
   self: grad * -((-self * self + 1).rsqrt())
+  result: element_wise
```
So I understand what `element_wise` and `linear` are from chatting with you. I don't see any documentation of these around (but I also haven't looked hard enough into the rest of the stack). It would be nice to add documentation for these, especially with an explanation for what they do

albanD replied:
Yes, I will add an updated version of the issue description here!

albanD replied:
And in derivatives.yaml

albanD committed:
Update on "Add many forward grad"




[ghstack-poisoned]

albanD committed:
Update on "Add many forward grad"




[ghstack-poisoned]

albanD committed:
Update on "Add many forward grad"




[ghstack-poisoned]

albanD committed:
Update on "Add many forward grad"




[ghstack-poisoned]

albanD committed:
Update on "Add many forward grad"




[ghstack-poisoned]

albanD committed:
Update on "Add many forward grad"




[ghstack-poisoned]

albanD committed:
Update on "Add many forward grad"




[ghstack-poisoned]

albanD committed:
Update on "Add many forward grad"




[ghstack-poisoned]

albanD committed:
Update on "Add many forward grad"




[ghstack-poisoned]

albanD committed:
Update on "Add many forward grad"


This PR add some helper to generate forward mode formula automatically based on the forward and backward formula.
The special case that are interesting to re-use are:
- if the function is element-wise (so there is a 1-1 mapping between the elements of the input and output), then the Jacobian is diagonal. And in this case, J v and v^T J are actually the same thing. So we can just re-use the backward function as-is, just providing a different v.
- if the function is linear (not affine!) of a single input (we might be able to do more than one, but that’s tricky) then f(x) = A x for some matrix A. And the Jacobian of f is also A. So we can do J v = A v = f(v). So we can just call the original function, replacing the differentiable input with the vector v of interest.



[ghstack-poisoned]

albanD committed:
Update on "Add many forward grad"


This PR add some helper to generate forward mode formula automatically based on the forward and backward formula.
The special case that are interesting to re-use are:
- if the function is element-wise (so there is a 1-1 mapping between the elements of the input and output), then the Jacobian is diagonal. And in this case, J v and v^T J are actually the same thing. So we can just re-use the backward function as-is, just providing a different v.
- if the function is linear (not affine!) of a single input (we might be able to do more than one, but that’s tricky) then f(x) = A x for some matrix A. And the Jacobian of f is also A. So we can do J v = A v = f(v). So we can just call the original function, replacing the differentiable input with the vector v of interest.



[ghstack-poisoned]

albanD committed:
Update on "Add many forward grad"


This PR add some helper to generate forward mode formula automatically based on the forward and backward formula.
The special case that are interesting to re-use are:
- if the function is element-wise (so there is a 1-1 mapping between the elements of the input and output), then the Jacobian is diagonal. And in this case, J v and v^T J are actually the same thing. So we can just re-use the backward function as-is, just providing a different v.
- if the function is linear (not affine!) of a single input (we might be able to do more than one, but that’s tricky) then f(x) = A x for some matrix A. And the Jacobian of f is also A. So we can do J v = A v = f(v). So we can just call the original function, replacing the differentiable input with the vector v of interest.

Differential Revision: [D23454148](https://our.internmc.facebook.com/intern/diff/D23454148)

[ghstack-poisoned]
